enhancements/ingress/optional-ingress-hypershift.md

```diff
@@ -0,0 +1,320 @@
+---
+title: optional-ingress-hypershift
+authors:
+  - "@alebedev87"
+reviewers:
+  - "@Miciah"
+  - "@bparees"
+  - "@cewong"
+  - "@wking"
+approvers:
+  - "@Miciah"
+  - "@bparees"
+  - "@wking"
+api-approvers:
+  - N/A
+creation-date: 2023-06-02
+last-updated: 2023-06-02
+tracking-link:
+  - https://issues.redhat.com/browse/NE-1129
+see-also:
+  - "/enhancements/installer/component-selection.md"
+  - "/enhancements/ingress/aws-load-balancer-operator.md"
+replaces:
+superseded-by:
+---
+
+# Make Ingress Operator optional on HyperShift
+
+## Summary
+
+This enhancement describes the implementation of a new capability to enable/disable the default ingress of the OpenShift cluster.
+The new capability (from here on called the ingress capability) includes [the cluster ingress operator](https://github.com/openshift/cluster-ingress-operator) as well as the custom resources managed by the operator.
+This enhancement focuses on the requirements for [RFE-3395](https://issues.redhat.com/browse/RFE-3395) which targets the ROSA clusters spawned and managed by HyperShift. The implementation of the ingress capability on the standalone OpenShift is out of the scope of this enhancement.
+
+## Motivation
+
+Users of the ROSA managed OpenShift service would like to use [AWS load balancers](/enhancements/ingress/aws-load-balancer-operator.md) as a way to accept the ingress traffic into the cluster.
+The cluster ingress operator becomes unnecessary and can be disabled to:
+- reduce the resource consumption on
+  - the HyperShift client clusters (ingress controllers)
+  - the HyperShift hosted control plane (operator)
+- make the HyperShift hosted control plane lighter and simpler, therefore more manageable
+- enable the HyperShift cluster users to make their choice for the cluster ingress
+
+### Goals
+- Implement the new ingress capability.
+- Add the ingress capability to the ingress operator's payload.
+- Make the dependent cluster operators tolerate the absence of the ingress operator.
+- Do not lock the subsequent implementation for the standalone OpenShift.
+
+### Non-Goals
+- Use a different way for enabling/disabling the ingress other than via [the cluster capability](/enhancements/installer/component-selection.md).
+- Fully implement the ingress capability for the standalone OpenShift:
+  - Make the OpenShift installer tolerate the absence of the default ingress controller.
+  - Make the cluster authentication operator tolerate the absence of the default ingress controller.
+  - Make the cluster monitoring operator tolerate the absence of the default ingress controller.
+- Disable the Route API or try to tolerate the absence of the default ingress controller.
```

Comment:
i think in this context it means "don't report yourself as failing/degraded, and everything about the component should keep working except any aspect that directly depends on having a functioning Route or ingress controller"

Reply:
> "don't report yourself as failing/degraded, and everything about the component should keep working except any aspect that directly depends on having a functioning Route or ingress controller"

Yes.

Rephrasing this non goal to make this clearer.

Reply:
I find the revised phrasing confusing too:

> Tolerate the unadmitted routes for any workload except for the dependent cluster operators

Are you saying that once this EP is implemented, if a HyperShift has the ingress capability turned off, and if there are unadmitted routes, then it is fine not to "tolerate" those routes?  Does that mean it is expected that creating a route could break the cluster when the ingress capability were off?  

Put another way, if you made this non-goal into a goal, what would need to be changed to "tolerate" unadmitted routes?

Reply:
> Are you saying that once this EP is implemented, if a HyperShift has the ingress capability turned off, and if there are unadmitted routes, then it is fine not to "tolerate" those routes? 

No. It may be not "fine". What I meant is that it's out of the scope of this EP.

> Does that mean it is expected that creating a route could break the cluster when the ingress capability were off?

No, it's not expected that "creating a route" could break the cluster. The HyperShift cluster should be able to become ready (all the cluster operators should be `Ready` and not `Degraded`).
However, certain user workloads or addon operators may be at risk of not being exposed or potentially broken.

> Put another way, if you made this non-goal into a goal, what would need to be changed to "tolerate" unadmitted routes?

I think that the only real solution in that case would be to implement the OpenShift route support in ALBC.